### PR TITLE
Adding TPC post-processing task and example json to generate quality overview panel for shifter

### PR DESCRIPTION
**Pull request by mlesch**

### CODE REVIEW

**Comment by Barthelemy** [comment]
@wiechula do you want to review ?

File: Modules/TPC/include/TPC/QualityObserver.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ private:
+  /// \brief Method to get the qualities of the QOs
+  void getQualities(const Trigger& t, o2::quality_control::repository::DatabaseInterface&);
+  /// \brief Method to create and publish the overview panel
+  void generatePanel();
+
+  std::vector<Config> mConfig;
+  std::string mObserverName;
+  std::map<std::string, std::vector<std::string>> Qualities;
+  std::map<std::string, int> mColors;
```
Comment (wiechula) [inline Modules/TPC/include/TPC/QualityObserver.h:73]: `unordered_map` + Naming convention:
```suggestion
  std::unordered_map<std::string, std::vector<std::string>> mQualities;
  std::unordered_map<std::string, int> mColors;
```

File: Modules/TPC/include/TPC/QualityObserver.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+/// \file     QualityObserver.h
+/// \author   Marcel Lesch
+///
+
+#ifndef QUALITYCONTROL_QUALITYOBSERVER_H
+#define QUALITYCONTROL_QUALITYOBSERVER_H
+
+#include "QualityControl/PostProcessingInterface.h"
+
+#include <memory>
```
Comment (wiechula) [inline Modules/TPC/include/TPC/QualityObserver.h:22]: Needed?
```suggestion
```

File: Modules/TPC/include/TPC/QualityObserver.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+#include <unordered_map>
+#include <vector>
+#include <string>
+#include <TCanvas.h>
```
Comment (wiechula) [inline Modules/TPC/include/TPC/QualityObserver.h:26]: Move to cxx use forward declaration here:
```suggestion
class TCanvas;
```

File: Modules/TPC/src/QualityObserver.cxx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    mConfig.push_back(dataConfig);
+
+    inputQO.clear();
+    inputQOTitle.clear();
+  } // for (const auto& dataSourceConfig : config.get_child("qc.postprocessing." + name + ".ratioConfig"))
+}
+
+void QualityObserver::initialize(Trigger, framework::ServiceRegistry&)
+{
+  for (const auto config : mConfig) {
```
Comment (wiechula) [inline Modules/TPC/src/QualityObserver.cxx:69]:
```suggestion
  for (const auto& config : mConfig) {
```

File: Modules/TPC/src/QualityObserver.cxx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  generatePanel();
+}
+
+void QualityObserver::finalize(Trigger t, framework::ServiceRegistry&)
+{
+  generatePanel();
+  if (mCanvas) {
+    delete mCanvas;
+    mCanvas = nullptr;
+  }
```
Comment (wiechula) [inline Modules/TPC/src/QualityObserver.cxx:91]: Check not really needed.
```suggestion
    delete mCanvas;
    mCanvas = nullptr;
```

File: Modules/TPC/src/QualityObserver.cxx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+void QualityObserver::getQualities(const Trigger& t,
+                                   repository::DatabaseInterface& qcdb)
+{
+  for (const auto& config : mConfig) {
+
+    if (Qualities[config.GroupTitle].size() > 0) {
+      Qualities[config.GroupTitle].clear();
+    }
+    for (const std::string qualityobject : config.QO) {
```
Comment (wiechula) [inline Modules/TPC/src/QualityObserver.cxx:102]:
```suggestion
    for (const auto& qualityobject : config.qo) {
```

File: Modules/TPC/src/QualityObserver.cxx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+      auto qo = qcdb.retrieveQO(config.Path + "/" + qualityobject, t.timestamp, t.activity);
```
Comment (wiechula) [inline Modules/TPC/src/QualityObserver.cxx:103]:
```suggestion
      const auto qo = qcdb.retrieveQO(config.Path + "/" + qualityobject, t.timestamp, t.activity);
```

File: Modules/TPC/src/QualityObserver.cxx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+        if (qualitylevel == 1) {
+          Qualities[config.GroupTitle].push_back("good");
+        } else if (qualitylevel == 2) {
+          Qualities[config.GroupTitle].push_back("medium");
+        } else if (qualitylevel == 3) {
+          Qualities[config.GroupTitle].push_back("bad");
+        } else {
+          Qualities[config.GroupTitle].push_back("no quality");
+        }
```
Comment (wiechula) [inline Modules/TPC/src/QualityObserver.cxx:117]: Why don't you simply use the official definitions?
https://github.com/AliceO2Group/QualityControl/blob/master/Framework/src/Quality.cxx#L28
```suggestion
   mQualities[config.GroupTitle].push_back(quality.getName());
```

File: Modules/TPC/src/QualityObserver.cxx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+void QualityObserver::initialize(Trigger, framework::ServiceRegistry&)
+{
+  for (const auto config : mConfig) {
+    Qualities[config.GroupTitle] = std::vector<std::string>();
+  }
+  mColors["bad"] = kRed;
+  mColors["medium"] = kOrange - 3;
+  mColors["good"] = kGreen + 2;
+  mColors["no quality"] = kGray + 2;
```
Comment (wiechula) [inline Modules/TPC/src/QualityObserver.cxx:75]: Perhaps:
```suggestion
  mColors[Quality::Bad.getName()] = kRed;
  mColors[Quality::Medium.getName()] = kOrange - 3;
  mColors[Quality::Good.getName()] = kGreen + 2;
  mColors[Quality::Null.getName()] = kGray + 2;
```

File: Modules/TPC/src/QualityObserver.cxx [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+          Qualities[config.GroupTitle].push_back("good");
+        } else if (qualitylevel == 2) {
+          Qualities[config.GroupTitle].push_back("medium");
+        } else if (qualitylevel == 3) {
+          Qualities[config.GroupTitle].push_back("bad");
+        } else {
+          Qualities[config.GroupTitle].push_back("no quality");
+        }
+      } else {
+        Qualities[config.GroupTitle].push_back("no quality");
```
Comment (wiechula) [inline Modules/TPC/src/QualityObserver.cxx:119]:
```suggestion
        mQualities[config.GroupTitle].push_back(Quality::Null.getName());
```

**Review by wiechula — CHANGES_REQUESTED** [review]

File: Modules/TPC/include/TPC/QualityObserver.h (hunk 3 above)
Comment (mlesch) [reply Modules/TPC/include/TPC/QualityObserver.h]: Changed as requested

File: Modules/TPC/include/TPC/QualityObserver.h [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  void configure(std::string name, const boost::property_tree::ptree& config) final;
+  void initialize(Trigger, framework::ServiceRegistry&) final;
+  void update(Trigger, framework::ServiceRegistry&) final;
+  void finalize(Trigger, framework::ServiceRegistry&) final;
+
+  struct Config {
+    std::string GroupTitle;
+    std::string Path;
+    std::vector<std::string> QO;
+    std::vector<std::string> QOTitle;
```
[the remark this answers is not in the document]
Comment (mlesch) [reply Modules/TPC/include/TPC/QualityObserver.h]: Changed as requested

File: Modules/TPC/src/QualityObserver.cxx (hunk 6 above)
Comment (mlesch) [reply Modules/TPC/src/QualityObserver.cxx]: Changed as requested

File: Modules/TPC/src/QualityObserver.cxx (hunk 8 above)
Comment (mlesch) [reply Modules/TPC/src/QualityObserver.cxx]: Changed as requested

File: Modules/TPC/src/QualityObserver.cxx (hunk 10 above)
Comment (mlesch) [reply Modules/TPC/src/QualityObserver.cxx]: Changed as requested

File: Modules/TPC/src/QualityObserver.cxx (hunk 5 above)
Comment (mlesch) [reply Modules/TPC/src/QualityObserver.cxx]: Changed as requested

**Review by wiechula — APPROVED** [review]

**Comment by Barthelemy** [comment]
@wiechula I will do it, I had to fix a conflict thus I am waiting for a few tests to pass

**Pull request merged**